lang/cs/Org.Apache.REEF.Client/YARN/RESTClient/YarnConfigurationUrlProvider.cs

```diff
@@ -0,0 +1,116 @@
+        private static readonly string YarnConfigFileName = "yarn-site.xml";
+        private static readonly string YarnRmWebappHttpsAddressPropertyName = "yarn.resourcemanager.webapp.https.address";
+        private static readonly string YarnRmWebappHttpAddressPropertyName = "yarn.resourcemanager.webapp.address";
+        private static readonly Logger Logger = Logger.GetLogger(typeof(YarnConfigurationUrlProvider));
+        private Uri _yarnRmUri;
+
+        [Inject]
+        private YarnConfigurationUrlProvider(
+            [Parameter(typeof(UseHttpsForYarnCommunication))] bool useHttps)
+            : this(null, useHttps)
+        {
+        }
+
+        [Inject]
+        private YarnConfigurationUrlProvider(
+            [Parameter(typeof(HadoopConfigurationDirectory))] string hadoopConfigDir,
+            [Parameter(typeof(UseHttpsForYarnCommunication))] bool useHttps)
+        {
+            hadoopConfigDir = hadoopConfigDir ?? Environment.GetEnvironmentVariable(HadoopConfDirEnvVariable);
+
+            if (string.IsNullOrEmpty(hadoopConfigDir) || !Directory.Exists(hadoopConfigDir))
+            {
+                throw new ArgumentException(HadoopConfDirEnvVariable + " is not configured or does not exist.",
+                    "hadoopConfigDir");
+            }
+
+            Logger.Log(Level.Verbose, "Using {0} as hadoop configuration directory", hadoopConfigDir);
+            string yarnConfigurationFile = Path.Combine(hadoopConfigDir, YarnConfigFileName);
+            LoadYarnConfiguration(yarnConfigurationFile, useHttps);
+
+            var configFileWatcher = new FileSystemWatcher(hadoopConfigDir, YarnConfigFileName)
+            {
+                NotifyFilter = NotifyFilters.LastWrite,
+                EnableRaisingEvents = true
+            };
+
+            TestHook_ConfigReloaded = new ManualResetEventSlim(false);
+            configFileWatcher.Changed += (sender, args) =>
+            {
+                LoadYarnConfiguration(yarnConfigurationFile, useHttps);
+                TestHook_ConfigReloaded.Set();
+            };
+        }
+
+        internal ManualResetEventSlim TestHook_ConfigReloaded { get; set; }
+
+        public Task<Uri> GetUrlAsync()
+        {
+            return Task.FromResult(_yarnRmUri);
+        }
+
+        private void LoadYarnConfiguration(string yarnConfigurationFile, bool useHttps)
+        {
+            var configRoot = XElement.Load(yarnConfigurationFile);
+
+            string propertyName = useHttps
+                ? YarnRmWebappHttpsAddressPropertyName
+                : YarnRmWebappHttpAddressPropertyName;
+
+            string prefix = useHttps ? "https" : "http";
+
+            var address = (string)configRoot.Elements("property")
+                .Where(x =>
+                    (string)x.Element("name") == propertyName)
+                .Select(x => x.Element("value")).FirstOrDefault();
+
+            if (!address.EndsWith("/"))
```

Comment:
I prefer ``address.TrimEnd("/") + "/"``, though it's less efficient, it also guarantees that there's only one ``'/'`` at the end.